compiler/src/main/scala/org/scalajs/core/compiler/JSTreeExtractors.scala

```diff
@@ -14,23 +14,51 @@ import org.scalajs.core.ir.Types._
 object JSTreeExtractors {
 
   object jse {
+
     /**
-     *  A literally named sequence (like in a call to applyDynamicNamed)
+     *  A sequence of tuples (like in a call to applyDynamicNamed)
      *
      *  Example (Scala): method(("name1", x), ("name2", y))
      */
-    object LitNamed {
-      def unapply(exprs: List[Tree]) = unapply0(exprs, Nil)
-
+    private object GenLitNamed {
       @tailrec
-      private def unapply0(
+      final def unapply0(
           exprs: List[Tree],
-          acc: List[(StringLiteral, Tree)]
+          acc: List[(StringLiteral, Tree)],
+          failIfNonLit: Boolean
         ): Option[List[(StringLiteral, Tree)]] = exprs match {
         case Tuple2(name: StringLiteral, value) :: xs =>
-          unapply0(xs, (name, value) :: acc)
+          unapply0(xs, (name, value) :: acc, failIfNonLit)
+        case _ :: xs =>
+          if (failIfNonLit)
+            None
+          else
+            unapply0(xs, acc, failIfNonLit)
         case Nil => Some(acc.reverse)
-        case _   => None
+      }
+    }
+
+    /**
+     *  A literally named sequence (like in a call to applyDynamicNamed)
+     *  Where all parameters are expected to be literally named.
+     *
+     *  Example (Scala): method(("name1", x), ("name2", y))
+     */
+    object LitNamed {
+      def unapply(exprs: List[Tree]): Option[List[(StringLiteral, Tree)]] = {
+        GenLitNamed.unapply0(exprs, Nil, true)
+      }
+    }
+
+    /**
+     *  A partially literally named sequence (like in a call to applyDynamicNamed)
+     *  Where some parameters can expected to be literally named.
+     *
+     *  Example (Scala): method(("name1", x), (a, y), z)
+     */
+    object ExtractedLitNamed {
```

Comment:
I've changed it to a method and removed one of the additional objects that I was creating.